include/upstream_temperature_feedback.hxx

```diff
@@ -0,0 +1,173 @@
+                      {{"time_dimension", "t"},
+                      {"units", "W s^-1"},
+                    {"conversion", Pnorm * Omega_ci},
+                    {"standard_name", "temperature source"},
+                    {"long_name", name + "upstream temperature feedback controller source"},
+                    {"source", "upstream_temperature_feedback"}});
+
+      // Save proportional and integral component of source for diagnostics/tuning
+      // Multiplier = proportional term + integral term
+      set_with_attrs(
+          state[std::string("temperature_feedback_src_p_") + name], proportional_term,
+          {{"time_dimension", "t"},
+          {"long_name", name + " proportional feedback term"},
+          {"source", "upstream_temperature_feedback"}});
+
+      
+      set_with_attrs(
+          state[std::string("temperature_feedback_src_i_") + name], integral_term,
+          {{"time_dimension", "t"},
+           {"long_name", name + " integral feedback term"},
+           {"source", "upstream_temperature_feedback"}});
+
+    }
+  }
+
+  void restartVars(Options& state) override {
+    AUTO_TRACE();
+
+    // NOTE: This is a hack because we know that the loaded restart file
+    //       is passed into restartVars in PhysicsModel::postInit
+    // The restart value should be used in init() rather than here
+    static bool first = true;
+    if (first and state.isSet(name + "_temperature_error_integral")) {
+      first = false;
+      temperature_error_integral = state[name + "_temperature_error_integral"].as<BoutReal>();
+    }
+
+    // Save the temperature error integral
+    set_with_attrs(state[name + "_temperature_error_integral"], temperature_error_integral,
+                   {{"long_name", name + " temperature error integral"},
+                    {"source", "upstream_temperature_feedback"}});
+  }
+
+private:
+  std::string name; ///< The species name
+
+  BoutReal temperature_upstream;                           ///< Normalised upstream temperature
+  BoutReal temperature_controller_p, temperature_controller_i; ///< PI controller parameters
+  BoutReal error;
+  BoutReal temperature_error_integral{0.0}; ///< Time integral of the error
+
+  bool temperature_integral_positive; ///< Force integral term to be positive?
+  bool temperature_source_positive;   ///< Force source to be positive?
+
+  // Terms used in Trapezium rule integration of error
+  BoutReal temperature_error_lasttime{-1.0};
+  BoutReal temperature_error_last{0.0};
+
+  Field3D pressure_source_shape; ///< This shape source is scaled up and down
+
+  BoutReal source_multiplier; ///< Factor to multiply source
+
+  BoutReal proportional_term, integral_term; ///< Components of resulting source for diagnostics
+
+  bool diagnose; ///< Output diagnostic information?
+};
+
+namespace {
+RegisterComponent<UpstreamTemperatureFeedback> register_uds("upstream_temperature_feedback");
```

Comment:
The name "register_uds" conflicts with the definition in `upstream_density_feedback`, when the headers are included together. Just change to "register_uts" or other unique name.

Reply:
Thanks! It's compiling now